electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,116 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+    super().__init__()
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_in = self.Port(VoltageSink.empty())
+    self.pwr_out = self.Port(VoltageSource.empty())
+
+    self.gate_resistor = self.ArgParameter(gate_resistor)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    output_current_draw = self.pwr_out.link().current_drawn
+    self.fet = self.Block(Fet.PFet(
+      drain_voltage=(0, self.pwr_out.link().voltage.upper()),
+      drain_current=output_current_draw,
+      gate_voltage=(- self.pwr_out.link().voltage.upper(), self.pwr_out.link().voltage.upper()),
+      rds_on=self.rds_on,
+    ))
+
+    self.res = self.Block(Resistor(self.gate_resistor))
+    # TODO: generate zener diode for high voltage applications
+    #  self.diode = self.Block(ZenerDiode(self.clamp_voltage))
+
+    self.import_kicad(
+      self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+      conversions={
+        'pwr_in': VoltageSink(
+          current_draw=output_current_draw,
+        ),
+        'pwr_out': VoltageSource(
+          voltage_out=self.pwr_in.link().voltage,
+        ),
+        'gnd': Ground(),
+      })
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  """Charging capable a battery reverse protection using PMOS transistors. The highest battery voltage is bounded by the
+  transistors' Vgs/Vds. There is also a rare case when this circuit being disconnected when a charger is connected first.
+  But always reverse protect. R1 and R2 are the pullup bias resistors for mp1 and mp2 PFet.
+  More info at: https://www.edn.com/reverse-voltage-protection-for-battery-chargers/
+  """
+
+  @init_in_parent
+  def __init__(self, r1_val: RangeLike = 100 * kOhm(tol=0.01), r2_val: RangeLike = 100 * kOhm(tol=0.01),
+               rds_on: RangeLike = (0, 0.1) * Ohm):
+    super().__init__()
+
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_out = self.Port(VoltageSource.empty(), doc="Power output for a load which will be also reverse protected from the battery")
+    self.pwr_in = self.Port(VoltageSink.empty(), doc="Power input from the battery")
+    self.chg_in = self.Port(VoltageSink.empty(), doc="Charger input to charge the battery")
+    self.chg_out = self.Port(VoltageSource.empty(), doc="Charging output to the battery chg port")
+
+    self.r1_val = self.ArgParameter(r1_val)
+    self.r2_val = self.ArgParameter(r2_val)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    self.r1 = self.Block(Resistor(resistance=self.r1_val))
+    self.r2 = self.Block(Resistor(resistance=self.r2_val))
+
+    batt_voltage = self.pwr_in.link().voltage.hull(self.chg_in.link().voltage).hull(self.gnd.link().voltage)
+
+    # taking the max of the current for the both direction. 0 lower bound
+    batt_current = self.pwr_in.link().current_limits.hull(self.chg_out.link().current_drawn).hull((0, 0))
```

Comment:
```suggestion
    batt_current = self.pwr_out.link().current_drawn.hull(self.chg_out.link().current_drawn).hull((0, 0))
```

Use current drawn instead of limits / capabilities?

Reply:
applied in be8264d